codeU4.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Sat Jun 10 09:17:26 2017
+Exercise 4, version 1
+@author: Luiza
+"""
+
+import unittest
+import numpy as np
+
+class Tile(object):
+
+    '''Given (n by m) bool 2D array consisting of islands as True
+    and water as False
+    Counts the number of separate islands, that can be formed using 4-connectivity
+    between the land tiles
+    '''
+
+    def __init__(self, bool_matrix, rows, cols):
+
+        '''Initialization of tile as 2D numpy bool array (True/False or 0/1)
+        checks whether we have the correct size and expects boolean array elements
+
+        Args:
+            bool_matrix: boolean numpy 2D array for tile
+            rows, cols: size of the tile array
+        '''
+
+        (self.rows, self.cols) = np.shape(bool_matrix)
+        self.tile = bool_matrix
+        assert self.rows == rows or self.cols == cols, \
+        "Given matrix dimensions mismatch between the actual size of array"
+        self.num_islands = 0
+        self.unvisited_land_cells = set((i, j) for i in range(self.rows) \
+        for j in range(self.cols) if self.tile[i, j])
+
+    def is_unvisited_island(self, tuple_cell):
+
+        """Checks if a given cell (x,y) is a land and was not visited so far
+
+        Args:
+            tuple_cell: a tuple (x,y) coordinate
+
+        Returns:
+            True: if not visited land
+            False: otherwise
+        """
+
+        return (0 <= tuple_cell[0] < self.rows and 0 <= tuple_cell[1] < self.cols \
+        and (tuple_cell[0], tuple_cell[1]) in self.unvisited_land_cells)
+
```

Comment:
I would just make these two separate checks. Packing the `self.unvisited_land_cells` check and bounds check into the same statement is not necessary and I think makes this a bit harder to follow.

Checking bounds with a tuple sounds like a clever trick. It might have the tiniest performance overhead (since it does have to pack and unpack things a bit behind the scenes) but for 2-tuples that should be negligible.

But oh, how does tuple comparison work in Python? I think it is not quite an `and` check - it checks elements in sequence and keeps on going only while the values are equal:

```
>>> (0, 0) <= (1, 1)
True
>>> (0, 0) <= (-1, 1)
False
>>> (0, 0) <= (1, -1)
True
```